### PR TITLE
Support for the case of `mpz`

### PR DESCRIPTION
**Pull request by haru-44**
#### References to other Issues or PRs

https://github.com/sympy/sympy/issues/25477#issuecomment-1669489565

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
Can you add ntheory here and in the other list as well:
https://github.com/sympy/sympy/blob/b68285262a06a914d9d920aba980d726be2efc39/bin/test_optional_dependencies.py#L44-L45

I guess that is why this was not previously noticed in CI.

**Comment by haru-44** [comment]
Is this correct?

File: sympy/ntheory/ecm.py [hunk 1]
```diff
@@ -341,7 +341,7 @@ def ecm(n, B1=10000, B2=100000, max_curve=200, seed=1234):
                 n //= prime
     while(n > 1):
         try:
-            factor = _ecm_one_factor(n, B1, B2, max_curve, seed + n)
+            factor = _ecm_one_factor(n, B1, B2, max_curve, seed)
```
Comment (smichr) [inline sympy/ntheory/ecm.py:344]: I saw this on the first review and thought it made sense, but now I think this underscores the wrongness of it. Could you look at the intro to sympy/core/random.py and see if you could convert this to use randint and seed as described there?

**Comment by oscarbenjamin** [comment]
> Is this correct?

Yes, I think so. We can see now that it exposes some ntheory failures:
```
________________________________________________________________________________
______________________ sympy.ntheory.generate.compositepi ______________________
File "/home/runner/work/sympy/sympy/sympy/ntheory/generate.py", line 1084, in sympy.ntheory.generate.compositepi
Failed example:
    compositepi(1000)
Expected:
    831
Got:
    830
________________________________________________________________________________
_______________________ sympy.ntheory.generate.composite _______________________
File "/home/runner/work/sympy/sympy/sympy/ntheory/generate.py", line 1021, in sympy.ntheory.generate.composite
Failed example:
    composite(17737)
Expected:
    20000
Got:
    20001
```
Are those related to the changes here or something from before?

**Comment by haru-44** [comment]
A mistake was made in the fix in #25204.
When `sieve.extend` is called, `sieve._list` may not be updated as expected because it calls `sieve.primerange`.

The following code fails.

```python
from sympy.ntheory.generate import sieve

sieve._reset(prime=True)
sieve.extend(100000)
assert len(sieve._list) == len(set(sieve._list))
```

File: sympy/ntheory/ecm.py (hunk 1 above)
Comment (haru-44) [reply sympy/ntheory/ecm.py:344]: Is it something like this?

```python
from sympy.core.random import randint, seed as seed_func

def _ecm_one_factor(n, B1=10000, B2=100000, max_curve=200, seed=None):
    seed_func(seed)
    ...
    sigma = randint(6, n - 1)
```

This way, the behavior is different when `seed=None`. However, it is not an important difference in `ecm`.

Do you want to fix the other places that use `_randint` as well? If so, we will need to consider them individually.

**Comment by oscarbenjamin** [comment]
I don't know ntheory so well so I will let others review this.

Looking at the doctest failure on master though it manifests as:
```
    raise ValueError("Increase the bounds")
ValueError: Increase the bounds
```
That comes from here:
https://github.com/sympy/sympy/blob/68f7e71c5c22948c1423589ba179c5080a596768/sympy/ntheory/ecm.py#L343-L346
This is an excellent example of bad error handling.

Firstly the expectation that it expects to catch is this one:
https://github.com/sympy/sympy/blob/68f7e71c5c22948c1423589ba179c5080a596768/sympy/ntheory/ecm.py#L304-L305
So the code catches the error only to re-raise the same error which is pointless.

Secondly the code actually might catch an unexpected error: something else might raise ValueError (is that what happens with gmpy2?). If there is any reason to catch the error then the *same* exception that was caught should be raised. The original exception has the most detailed information about the cause of the problem and should not be supressed.

Thirdly there is no point in `_ecm_one_factor` raising a `ValueError` as the last line of the function and then the calling function catching that error. Instead `_ecm_one_factor` could return `None` and the calling code could check for `None`.

Fourthly ValueError is just too generic an exception to *ever* be used as part of normal control flow. All kinds of bugs can cause a `ValueError` to be raised so it should only ever be caught around a *single* call to an external (not in the SymPy codebase) function that is known to raise ValueError. Within the SymPy codebase if anyone *ever* wants to catch a `ValueError` then the exception that is *raised* should be changed to something more specific like `ECMNoFactorFoundError` or `ECMBoundTooLow` or something.

In general there are far too many places in SymPy that raise generic exceptions like ValueError and TypeError rather than something more specific. Worse there are far too many places that *catch* these exceptions and those should all be removed.

**Comment by oscarbenjamin** [comment]
I am seeing some of the same ntheory failures in gh-25492.

I don't quite understand how gmpy2 (or python-flint) causes these tests to fail in the way that it does. It seems like something else should be fixed somewhere so that whatever problem arises with gmpy2 manifests as an exception rather than giving wrong results.

Is the problem caused by catching an exception somewhere?

**Comment by oscarbenjamin** [comment]
> ```python
> from sympy.ntheory.generate import sieve
> 
> sieve._reset(prime=True)
> sieve.extend(100000)
> assert len(sieve._list) == len(set(sieve._list))
> ```

I see that bug and that it is fixed here but that one shows up regardless of gmpy2:
```console
$ SYMPY_GROUND_TYPES=python python t.py
Traceback (most recent call last):
  File "/Users/enojb/current/sympy/sympy/t.py", line 5, in <module>
    assert len(sieve._list) == len(set(sieve._list))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
$ SYMPY_GROUND_TYPES=gmpy2 python t.py
Traceback (most recent call last):
  File "/Users/enojb/current/sympy/sympy/t.py", line 5, in <module>
    assert len(sieve._list) == len(set(sieve._list))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```
What I don't understand is this difference:
```console
$ SYMPY_GROUND_TYPES=gmpy2 bin/doctest sympy/ntheory
=========================================================================================== test process starts ============================================================================================
executable:         /Users/enojb/.pyenv/versions/sympy.git.3.11/bin/python  (3.11.3-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       gmpy 2.1.5
numpy:              1.24.2
hash randomization: on (PYTHONHASHSEED=3016577396)

sympy/ntheory/bbp_pi.py[1] .                                                                                                                                                                            [OK]
sympy/ntheory/continued_fraction.py[5] .....                                                                                                                                                            [OK]
sympy/ntheory/digits.py[3] ...                                                                                                                                                                          [OK]
sympy/ntheory/ecm.py[4] F...                                                                                                                                                                          [FAIL]
sympy/ntheory/egyptian_fraction.py[1] .                                                                                                                                                                 [OK]
sympy/ntheory/elliptic_curve.py[8] ........                                                                                                                                                             [OK]
sympy/ntheory/factor_.py[37] .....................................                                                                                                                                      [OK]
sympy/ntheory/generate.py[18] FF................                                                                                                                                                      [FAIL]
sympy/ntheory/modular.py[5] ..F..                                                                                                                                                                     [FAIL]
sympy/ntheory/multinomial.py[4] ....                                                                                                                                                                    [OK]
sympy/ntheory/partitions_.py[1] .                                                                                                                                                                       [OK]
sympy/ntheory/primetest.py[10] ..........                                                                                                                                                               [OK]
sympy/ntheory/qs.py[2] ..                                                                                                                                                                               [OK]
sympy/ntheory/residue_ntheory.py[26] ..........................                                                                                                                                         [OK]

____________________________________________________________________________________________________________________________________________________________________________________________________________
__________________________________________________________________________________________ sympy.ntheory.ecm.ecm ___________________________________________________________________________________________
File "/Users/enojb/current/sympy/sympy/sympy/ntheory/ecm.py", line 328, in sympy.ntheory.ecm.ecm
Failed example:
    ecm(25645121643901801)
Exception raised:
    Traceback (most recent call last):
      File "/Users/enojb/current/sympy/sympy/sympy/ntheory/ecm.py", line 344, in ecm
        factor = _ecm_one_factor(n, B1, B2, max_curve, seed + n)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/enojb/current/sympy/sympy/sympy/ntheory/ecm.py", line 216, in _ecm_one_factor
        randint = _randint(seed)
                  ^^^^^^^^^^^^^^
      File "/Users/enojb/current/sympy/sympy/sympy/core/random.py", line 227, in _randint
        raise ValueError('_randint got an unexpected seed')
    ValueError: _randint got an unexpected seed

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/Users/enojb/.pyenv/versions/3.11.3/lib/python3.11/doctest.py", line 1351, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest sympy.ntheory.ecm.ecm[1]>", line 1, in <module>
        ecm(25645121643901801)
      File "/Users/enojb/current/sympy/sympy/sympy/ntheory/ecm.py", line 346, in ecm
        raise ValueError("Increase the bounds")
    ValueError: Increase the bounds
**********************************************************************
File "/Users/enojb/current/sympy/sympy/sympy/ntheory/ecm.py", line 330, in sympy.ntheory.ecm.ecm
Failed example:
    ecm(9804659461513846513)
Exception raised:
    Traceback (most recent call last):
      File "/Users/enojb/current/sympy/sympy/sympy/ntheory/ecm.py", line 344, in ecm
        factor = _ecm_one_factor(n, B1, B2, max_curve, seed + n)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/enojb/current/sympy/sympy/sympy/ntheory/ecm.py", line 216, in _ecm_one_factor
        randint = _randint(seed)
                  ^^^^^^^^^^^^^^
      File "/Users/enojb/current/sympy/sympy/sympy/core/random.py", line 227, in _randint
        raise ValueError('_randint got an unexpected seed')
    ValueError: _randint got an unexpected seed

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/Users/enojb/.pyenv/versions/3.11.3/lib/python3.11/doctest.py", line 1351, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest sympy.ntheory.ecm.ecm[2]>", line 1, in <module>
        ecm(9804659461513846513)
      File "/Users/enojb/current/sympy/sympy/sympy/ntheory/ecm.py", line 346, in ecm
        raise ValueError("Increase the bounds")
    ValueError: Increase the bounds
____________________________________________________________________________________________________________________________________________________________________________________________________________
____________________________________________________________________________________ sympy.ntheory.generate.compositepi ____________________________________________________________________________________
File "/Users/enojb/current/sympy/sympy/sympy/ntheory/generate.py", line 1084, in sympy.ntheory.generate.compositepi
Failed example:
    compositepi(1000)
Expected:
    831
Got:
    830
____________________________________________________________________________________________________________________________________________________________________________________________________________
_____________________________________________________________________________________ sympy.ntheory.generate.composite _____________________________________________________________________________________
File "/Users/enojb/current/sympy/sympy/sympy/ntheory/generate.py", line 1021, in sympy.ntheory.generate.composite
Failed example:
    composite(17737)
Expected:
    20000
Got:
    20001
____________________________________________________________________________________________________________________________________________________________________________________________________________
________________________________________________________________________________________ sympy.ntheory.modular.crt1 ________________________________________________________________________________________
File "/Users/enojb/current/sympy/sympy/sympy/ntheory/modular.py", line 127, in sympy.ntheory.modular.crt1
Failed example:
    for v in vs:
        print(crt2(m, v, mm, e, s))
Expected:
    (397042, 912285)
    (803206, 912285)
Got:
    (mpz(397042), mpz(912285))
    (mpz(803206), mpz(912285))

========================================================================== tests finished: 121 passed, 4 failed, in 6.09 seconds ===========================================================================
DO *NOT* COMMIT!
$ SYMPY_GROUND_TYPES=python bin/doctest sympy/ntheory
=========================================================================================== test process starts ============================================================================================
executable:         /Users/enojb/.pyenv/versions/sympy.git.3.11/bin/python  (3.11.3-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python
numpy:              1.24.2
hash randomization: on (PYTHONHASHSEED=374551107)

sympy/ntheory/bbp_pi.py[1] .                                                                                                                                                                            [OK]
sympy/ntheory/continued_fraction.py[5] .....                                                                                                                                                            [OK]
sympy/ntheory/digits.py[3] ...                                                                                                                                                                          [OK]
sympy/ntheory/ecm.py[4] ....                                                                                                                                                                            [OK]
sympy/ntheory/egyptian_fraction.py[1] .                                                                                                                                                                 [OK]
sympy/ntheory/elliptic_curve.py[8] ........                                                                                                                                                             [OK]
sympy/ntheory/factor_.py[37] .....................................                                                                                                                                      [OK]
sympy/ntheory/generate.py[18] ..................                                                                                                                                                        [OK]
sympy/ntheory/modular.py[5] .....                                                                                                                                                                       [OK]
sympy/ntheory/multinomial.py[4] ....                                                                                                                                                                    [OK]
sympy/ntheory/partitions_.py[1] .                                                                                                                                                                       [OK]
sympy/ntheory/primetest.py[10] ..........                                                                                                                                                               [OK]
sympy/ntheory/qs.py[2] ..                                                                                                                                                                               [OK]
sympy/ntheory/residue_ntheory.py[26] ..........................                                                                                                                                         [OK]

=============================================================================== tests finished: 125 passed, in 6.11 seconds ================================================================================
```
The `ecm` failure makes sense but why do the other examples fail with gmpy2 and not with python?

**Comment by haru-44** [comment]
> A mistake was made in the fix in #25204.
> When `sieve.extend` is called, `sieve._list` may not be updated as expected because it calls `sieve.primerange`.

This bug is not affected by the python/gmpy difference. `compositepi` and others use `sieve` internally, and the state of `sieve` depends on the history of calls to `extend`.

When the following code is executed, `sieve.extend(100000)` is executed internally and the bug is exposed. The test case did not execute `sieve.extend(100000)`, so it could not be detected.

https://github.com/sympy/sympy/blob/68f7e71c5c22948c1423589ba179c5080a596768/doc/src/modules/ntheory.rst?plain=1#L226-L227

Are all your questions answered?

**Comment by oscarbenjamin** [comment]
> This bug is not affected by the python/gmpy difference. `compositepi` and others use `sieve` internally, and the state of `sieve` depends on the history of calls to `extend`.

Somehow the test failure depends on the python/gmpy diference so is it that gmpy affects the history of calls to extend somehow? I guess because somewhere gmpy is used instead of the sieve or something?

**Comment by haru-44** [comment]
I believe the difference is whether or not to run the `.rst` file test. do you test for `.rst` files in your python tests?

**Comment by oscarbenjamin** [comment]
> I believe the difference is whether or not to run the `.rst` file test.

The example I showed above does not run the .rst files. It is running the module doctests (same tests in both cases).

**Comment by haru-44** [comment]
I see. Then I have no idea.

**Comment by oscarbenjamin** [comment]
I was imagining something like this might affect it somehow:
https://github.com/sympy/sympy/blob/68f7e71c5c22948c1423589ba179c5080a596768/sympy/ntheory/primetest.py#L621-L625

**Comment by oscarbenjamin** [comment]
Okay, well I understand this PR now well enough to say let's get it in. The changes to random handling can be in a followup maybe but this fixes a clear bug and the doctests.

**Comment by haru-44** [comment]
Thanks.

The error depends on how `sieve.extend` is called, but in gmpy's case, the following code calls `sieve.extend` for the first time.

https://github.com/sympy/sympy/blob/b7a7ee1b7096486d8ecd460332506e04033cce13/sympy/ntheory/ecm.py#L328

In contrast, the python one calls `sieve.extend` for the first time in the following place.

https://github.com/sympy/sympy/blob/b7a7ee1b7096486d8ecd460332506e04033cce13/sympy/ntheory/digits.py#L103

In either python or gmpy, `digits.py` should have been executed first, so the question is why `sieve.primerange` was called in `digits.py` only in python.

So, if we check the code in `factorials.py`, we see that it uses gmpy.

https://github.com/sympy/sympy/blob/b7a7ee1b7096486d8ecd460332506e04033cce13/sympy/functions/combinatorial/factorials.py#L170-L176

**Comment by oscarbenjamin** [comment]
Okay, that makes sense. Thanks!

As noted in `factorials.py` that code should probably be moved out to somewhere else. Ideally we should minimise the number of places that have any conditional logic around gmpy (especially if we will also have python-flint as an alternative to gmpy). There should be one place where we figure out the best algorithms and implementations for each operation and the rest of the codebase should just depend on that place.

**Pull request merged**